DOCS-8034-vm-publishlisten-example-kt (#1039)
mulesoft/docs-connectors: vm/2.0/modules/ROOT/pages/vm-publish-listen.adoc

:page-aliases: connectors::vm-consume.adoc, connectors::vm/vm-publish-listen.adoc

> The following example illustrates how to publish a message that is received by a different flow. The first flow takes an incoming HTTP message payload, then the *VM Publish* operation publishes the message to a VM queue. In the second flow, the *VM Listener* source receives the message payload from the previous flow via the same VM queue. To do this, point both operations to the same queue which prevents message loss caused by sending messages to a queue that nobody is listening to.

Reviewer: ```suggestion
The following example illustrates how to publish a message that is received by a different flow:

* The first flow receives an incoming HTTP message payload, then the *VM Publish* operation publishes the message to a VM queue. 
* In the second flow, the *VM Listener* source receives the message payload from the previous flow using the same VM queue. +
To do this, point both operations to the same queue, which prevents message loss caused by sending messages to a queue that nobody is listening to.
```

Reviewer: I don't know about saying "nobody is listening to." Maybe something like "...a queue with no active listener?"

The following screenshot shows the Anypoint Studio app flow for this example:

.VM Connector Publish and Listen to Messages flow
image::vm-example-publishlisten.png[VM Connector Publish and Listen to Messages flow]

== Create the VM Publishing Flow

> The VM publishing flow first listens for an incoming HTTP message payload which contains the body attribute of a JSON object, then in a one-way operation the *VM Publish*  publishes the message in a VM queue, and the flow continues.

Reviewer: ```suggestion
The VM Connector publishing flow first listens for an incoming HTTP message payload, which contains the body attribute of a JSON object. Then, in a one-way *VM Publish* operation, the message is published in a VM queue, and the flow continues.
```

Reviewer: Maybe it is just VM publishing flow.  I'm not sure about that one. You know better so disregard if it's correct.

To create the VM publishing flow: +

. Create a new Mule project in Studio.

> . In the *Mule Palette* view, select the *HTTP Listener* source and drag it into the canvas. +

Reviewer: ```suggestion
. In the *Mule Palette* view, select the *HTTP Listener* source and drag it on to the canvas. +
```

> The source listens for incoming HTTP messages attributes.

Reviewer: I think you can just say the source listens for incoming messages... If not, then: 
The source listens for incoming HTTP message attributes.

. On the *HTTP Listener* configuration screen, set the *Path* field to `/json-here`.
. Click the plus sign (+) next to the *Connector configuration* field to configure a global element that can be used by all instances of the *HTTP Listener* in the app.

> . In the *Connection* section, set *Protocol* to `HTTP (Default)`, *Host* to `All Interfaces [0.0.0.0] (default)`, and *Port* to `8081`.

Reviewer: ```suggestion
. In the *Connection* section, enter the following values:
* *Protocol* +`
HTTP (Default)`
* *Host* 
`All Interfaces [0.0.0.0] (default)`
* *Port* 
`8081`
```

. Click *Test Connection* to confirm that Mule can connect with the specified server.
. Click *OK*.
. Drag the *VM Publish* operation to the right of the *HTTP Listener* source. +
The operation publishes the message payload in the VM queue.
. In the *Content* field of the connector configuration, set the value to `payload.body`. +

> This field enables you to use DataWeave to build the value to send, if you do not specify a parameter value, the entire payload is sent.

Reviewer: ```suggestion
This field enables you to use DataWeave to build the value to send. If you do not specify a parameter value, the entire payload is sent.
```

. On the *VM Publish* configuration screen, click the plus sign (+) next to the *Connector configuration* field to configure a global element for the VM queue.

> . On the *VM Config* window, for *Queues* select `Edit inline`.

Reviewer: ```suggestion
. On the *VM Config* window, for *Queues*, select *Edit inline*.
```

> . On the *Queue* window, set *Queue name* to `myQueue`, *Queue type* to `TRANSIENT (Default)`, and *Max outstanding message* to `o`.

Reviewer: ```suggestion
. In the *Queue* window, enter the following values:
* *Queue name* +
`myQueue`
* *Queue type* + 
`TRANSIENT (Default)`
* *Max outstanding message* +
`0`.
```
